ambry-frontend/src/main/java/com/github/ambry/frontend/NamedBlobsCleanupRunner.java

```diff
@@ -32,36 +37,61 @@ public class NamedBlobsCleanupRunner implements Runnable {
+  public NamedBlobsCleanupRunner(Router router, NamedBlobDb namedBlobDb, AccountService accountService) {
     this.router = router;
     this.namedBlobDb = namedBlobDb;
+    this.accountService = accountService;
   }
 
   @Override
   public void run() {
     logger.info("Named Blobs Cleanup Runner is initiated");
     try {
-      List<StaleNamedBlob> staleResultList = namedBlobDb.pullStaleBlobs().get();
-      List<StaleNamedBlob> failedResults = new ArrayList<>();
-      for (StaleNamedBlob staleResult : staleResultList) {
-        try {
-          router.deleteBlob(staleResult.getBlobId(), "ambry-named-blobs-cleanup-runner").get();
-        } catch (Exception e) {
-          if (!e.getMessage().contains(RouterErrorCode.BlobDoesNotExist.name())) {
-            logger.error("Failed to cleanup named stale blob {}", staleResult, e);
-            failedResults.add(staleResult);
-          }
+      Set<Container> activeContainers = accountService.getContainersByStatus(Container.ContainerStatus.ACTIVE);
+      Set<Container> inactiveContainers = accountService.getContainersByStatus(Container.ContainerStatus.INACTIVE);
+      Set<Container> combinedContainers = new HashSet<>(activeContainers);
+      combinedContainers.addAll(inactiveContainers);
+      List<StaleNamedBlob> batchStaleBlobs = Collections.emptyList();
+      NamedBlobDb.StaleBlobsWithLatestBlobName staleBlobsWithLatestBlobName;
+      for (Container container : combinedContainers) {
+        if (container.getNamedBlobMode() == Container.NamedBlobMode.DISABLED) {
+          continue;
         }
+        // set blobName to be "\0" since it is the lowest ASCII value and everything is greater than it
+        String blobName = smallestASCII;
+        do {
+          staleBlobsWithLatestBlobName = namedBlobDb.pullStaleBlobs(container, blobName).get();
+          batchStaleBlobs = staleBlobsWithLatestBlobName.getStaleBlobs();
+          List<StaleNamedBlob> failedResults = new ArrayList<>();
+          for (StaleNamedBlob staleBlob : staleBlobsWithLatestBlobName.getStaleBlobs()) {
+            try {
+              router.deleteBlob(staleBlob.getBlobId(), "ambry-named-blobs-cleanup-runner").get();
+            } catch (Exception e) {
+              if (!e.getMessage().contains(RouterErrorCode.BlobDoesNotExist.name())) {
+                logger.error("Failed to cleanup named stale blob {}", staleBlob, e);
+                failedResults.add(staleBlob);
+              }
+            }
+          }
+
+          batchStaleBlobs.removeAll(failedResults);
+          namedBlobDb.cleanupStaleData(batchStaleBlobs);
+
+          logger.info("Named Blobs Cleanup Runner processed {} stale blobs ({} failed deletions)",
+              batchStaleBlobs.size(), failedResults.size());
+
+          Set<String> cleanedBlobIds =
+              batchStaleBlobs.stream().map(StaleNamedBlob::getBlobId).collect(Collectors.toSet());
+          logger.info("The cleaned blobIds are: {}", cleanedBlobIds);
```

Comment:
you have to change blobName to staleBlobsWithLatestBlobName so we can continue the second page.